File: MessageU-Server/protocol_request.py
""" Request """
from enum import Enum

GENERAL_ERROR = 9000

class RequestCode(Enum):
    REGISTER_REQUEST = 100
    CLIENTS_LIST_REQUEST = 101
    PUBLIC_KEY_REQUEST = 102
    PUSH_MESSAGE_REQUEST = 103
    PULL_MESSAGES_REQUEST = 104

    @classmethod
    def is_valid(cls, value):
        return value in cls._value2member_map_


class Request:
    def __init__(self):
        self.header = RequestHeader()
        self.payload = None

    def set_suitable_payload(self):
        if self.header.code is not None:
            if self.header.code == RequestCode.REGISTER_REQUEST.value:
                self.payload = RegisterRequestPayload()
            elif self.header.code == RequestCode.PUBLIC_KEY_REQUEST.value:
                self.payload = PublicKeyRequestPayload()
            elif self.header.code == RequestCode.PUSH_MESSAGE_REQUEST.value:
                self.payload = MessagePayload()


class RequestHeader:
    def __init__(self):
        self.version = None
        self.code = None
        self.payload_size = None
        self.client_id = None

    @property
    def client_id(self):
        return self._client_id

    @client_id.setter
    def client_id(self, cid):
        self._client_id = cid

    @property
    def version(self):
        return self._version

    @version.setter
    def version(self, version):
        if version is None:
            self._code = None
        elif version in {1, 2}:
            self._version = version

    @property
    def code(self):
        return self._code

    @code.setter
    def code(self, code):
        if code is None:
            self._code = None
        if RequestCode.is_valid(code):
            self._code = code
        else:
            self._code = GENERAL_ERROR

    @property
    def payload_size(self):
        return self._payload_size

    @payload_size.setter
    def payload_size(self, payload_size):
        self._payload_size = payload_size


class RequestPayload:
    pass


class RegisterRequestPayload(RequestPayload):
    def __init__(self):
        self.name = None
        self.public_key = None

    @property
    def name(self):
        return self._name.split(b'\0', 1)[0].decode("utf-8")

    @name.setter
    def name(self, name):
        self._name = name

    @property
    def public_key(self):
        return self._public_key

    @public_key.setter
    def public_key(self, key):
        self._public_key = key


class PublicKeyRequestPayload(RequestPayload):
    def __init__(self):
        self.client_id = None

    @property
    def client_id(self):
        return self._client_id

    @client_id.setter
    def client_id(self, client_id):
        self._client_id = client_id


class MessagePayload(RequestPayload):
    def __init__(self):
        self.client_id = None
        self.message_type = None
        self.content_size = None
        self.message_content = None

    @property
    def client_id(self):
        return self._client_id

    @client_id.setter
    def client_id(self, client_id):
        self._client_id = client_id

    @property
    def message_type(self):
        return self._message_type

    @message_type.setter
    def message_type(self, message_type):
        self._message_type = message_type

    @property
    def content_size(self):
        return self._content_size

    @content_size.setter
    def content_size(self, content_size):
        self._content_size = content_size

    @property
    def message_content(self):
        return self._message_content

    @message_content.setter
    def message_content(self, message_content):
        self._message_content = message_content

File: MessageU-Client/protocol_request.cpp
/**
	Encode requests by the protocol specification.
	@file protocol.cpp
	@author Itay Matza
	@version 1.0 01/03/21
*/

#include "protocol_request.h"
using namespace std;
using boost::asio::ip::tcp;

// Writes to server. 
void writeToServer(tcp::socket& sock, uint8_t* request, unsigned long request_length) {
	size_t total_sent = 0;
	while (request_length > 0) {
		size_t size = min(static_cast<unsigned long>(CHUNK_SIZE), request_length);
		size_t sent = boost::asio::write(sock, boost::asio::buffer(request + total_sent, size));
		request_length -= sent;
	}
}

// Encode register request by the protocol specification.
RegisterRequest* encodeRegisterRequest(string username, uint8_t public_key[]) {
	RegisterRequest* request = new RegisterRequest;
	request->header.version = CLIENT_VERSION;
	request->header.code = RequestCode::REGISTER_REQUEST;
	request->header.payoad_size = sizeof(RegisterRequestPayload);

	size_t size = min(username.size(), (size_t) USERNAME_LEN - 1);
	copy(username.begin(), username.begin() + size, std::begin(request->payload.name));
	request->payload.name[size] = '\0';

	memcpy(request->payload.public_key, public_key, PUBKEY_LEN);

	return request;
}

// Encode client list request by the protocol specification.
ClientsListRequest* encodeClientsListRequest(uint8_t uid[UID_LEN]) {
	ClientsListRequest* request = new ClientsListRequest;
	memcpy(request->header.uid, uid, UID_LEN);
	request->header.version = CLIENT_VERSION;
	request->header.code = RequestCode::CLIENTS_LIST_REQUEST;
	request->header.payoad_size = 0;
	return request;
}

// Encode public key request by the protocol specification.
PublicKeyRequest* encodePublicKeyRequest(uint8_t uid[UID_LEN], uint8_t other_uid[UID_LEN]) {
	PublicKeyRequest* request = new PublicKeyRequest;
	memcpy(request->header.uid, uid, UID_LEN);
	request->header.version = CLIENT_VERSION;
	request->header.code = RequestCode::PUBLIC_KEY_REQUEST;
	request->header.payoad_size = sizeof(PublicKeyRequestPayload);
	memcpy(request->payload.uid, other_uid, UID_LEN);
	return request;
}

// Encode pull messages request by the protocol specification.
PullMessagesRequest* encodePullMessagesRequest(uint8_t uid[UID_LEN]){
	PullMessagesRequest* request = new PullMessagesRequest;
	memcpy(request->header.uid, uid, UID_LEN);
	request->header.version = CLIENT_VERSION;
	request->header.code = RequestCode::PULL_MESSAGES_REQUEST;
	request->header.payoad_size = 0;
	return request;
}

// Encode push message request by the protocol specification.
PushMessageRequest* encodePushMessageRequest(uint8_t uid[UID_LEN], uint8_t other_uid[UID_LEN], uint8_t message_type, size_t message_size) {
	PushMessageRequest* request = new PushMessageRequest;
	memcpy(request->header.uid, uid, UID_LEN);
	request->header.version = CLIENT_VERSION;
	request->header.code = RequestCode::PUSH_MESSAGE_REQUEST;
	request->header.payoad_size = sizeof(PushMessageRequestPayload) + message_size;
	memcpy(request->payload.uid, other_uid, UID_LEN);
	request->payload.message_type = message_type;
	request->payload.message_size = message_size;
	return request;
}